world/mujoco_ros/include/mujoco_ros/camera_publisher.hpp

```diff
@@ -0,0 +1,121 @@
+#pragma once
+#include "defs.h"
+#include <sstream>
+#include <cstdio>
+#include <mutex>
+#include <thread>
+
+#include <cv_bridge/cv_bridge.h>
+#include <opencv2/imgcodecs.hpp>
+
+#include <mujoco_ros/simulator.hpp>
+namespace mj_ros
+{
+class simulator_t;
+
+class camera_rgb_publisher_t
+{
+public:
+  camera_rgb_publisher_t(ros::NodeHandle& nh, SimulatorPtr sim, const std::string camera_name, double frequency = 30)
+    : _sim(sim)
+    , _rate(frequency)
+    , _root(ros::this_node::getNamespace())
+    , _topic_name(_root + "/camera/rgb")
+    , _publisher(nh.advertise<sensor_msgs::Image>(_topic_name, 1000))
+    , _cam_name(camera_name)
+    , _header()
+  {
+    mjv_defaultOption(&_mj_option);
+    mjv_defaultScene(&_mj_scene);
+    mjr_defaultContext(&_mj_context);
+
+    _header.seq = 0;
+    _header.stamp = ros::Time::now();
+    _header.frame_id = _topic_name;
+
+    if (!glfwInit())
+      ROS_ERROR("Error in initializing GLFW.");
+
+    glfwWindowHint(GLFW_VISIBLE, GLFW_FALSE);
+    _glfw_window = glfwCreateWindow(1280, 720, _cam_name.c_str(), NULL, NULL);
+
+    if (!_glfw_window)
+      ROS_ERROR("Error in creating GLFW window.");
+
+    glfwMakeContextCurrent(_glfw_window);
+    // The next three calls need to be AFTER glfwMakeContextCurrent
+    mjv_defaultCamera(&_mj_camera);
+    mjv_makeScene(_sim->m, &_mj_scene, 1000);
+    mjr_makeContext(_sim->m, &_mj_context, mjFONTSCALE_150);
+
+    for (int i = 0; i < _sim->m->ncam; ++i)
+    {
+      const std::string name_i{ _sim->m->names + _sim->m->name_camadr[i] };
+      // std::cout << "name_i: " << name_i << std::endl;
+      if (_cam_name == name_i)
+      {
+        _mj_viewport.left = 0;
+        _mj_viewport.bottom = 0;
+        // _mj_viewport.width = _sim->m->cam_resolution[i];
```

Comment:
I am a fan of this and not hardcoding it, but if there is a specific resolution that we want to use, this is fine.

Reply:
Agree with you, but there is a problem: `cam_resolution` is only on mj 3.0 and forward ([23](https://mujoco.readthedocs.io/en/stable/changelog.html#version-3-0-0-october-18-2023)). I'm currently using 2.7.

Reply:
Agreed, could you add a `TODO` for now (We can fix once we upgrade to 3.0)